[inductor] make requires_stride_order more unbacked-symint-aware

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #137064
* __->__ #137063

Previously, we tried to sort SymInt strides to determine the stride
order. This PR makes the sorting more unbacked symint aware: given a Tensor
with sizes (u0, u1, u2), it has strides (u1 * u2, u1, 1), which is
sortable under the guard_size_oblivious assumptions.

Test Plan:
- test case

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

zou3519 committed:
[inductor] make requires_stride_order more unbacked-symint-aware

Previously, we tried to sort SymInt strides to determine the stride
order. This PR makes the sorting more unbacked symint aware: given a Tensor
with sizes (u0, u1, u2), it has strides (u1 * u2, u1, 1), which is
sortable under the guard_size_oblivious assumptions.

Test Plan:
- test case

[ghstack-poisoned]

zou3519 committed:
Update on "[inductor] make requires_stride_order more unbacked-symint-aware"


Previously, we tried to sort SymInt strides to determine the stride
order. This PR makes the sorting more unbacked symint aware: given a Tensor
with sizes (u0, u1, u2), it has strides (u1 * u2, u1, 1), which is
sortable under the guard_size_oblivious assumptions.

Test Plan:
- test case

[ghstack-poisoned]

zou3519 committed:
Update on "[inductor] make requires_stride_order more unbacked-symint-aware"


Previously, we tried to sort SymInt strides to determine the stride
order. This PR makes the sorting more unbacked symint aware: given a Tensor
with sizes (u0, u1, u2), it has strides (u1 * u2, u1, 1), which is
sortable under the guard_size_oblivious assumptions.

Test Plan:
- test case

[ghstack-poisoned]

zou3519 committed:
Update on "[inductor] make requires_stride_order more unbacked-symint-aware"


Previously, we tried to sort SymInt strides to determine the stride
order. This PR makes the sorting more unbacked symint aware: given a Tensor
with sizes (u0, u1, u2), it has strides (u1 * u2, u1, 1), which is
sortable under the guard_size_oblivious assumptions.

Test Plan:
- test case

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

zou3519 commented on torch/_inductor/ir.py:
```diff
@@ -249,7 +250,11 @@ def get_stride_order(seq: Sequence[Union[int, torch.SymInt, Expr]]) -> Sequence[
     """
-    sorted_idx: List[int] = argsort(seq)
+    seq_lst = list(seq)
+    if all(isinstance(elt, (int, torch.SymInt)) for elt in seq_lst):
+        sorted_idx = argsort_symint(seq_lst)
+    else:
+        sorted_idx = argsort(seq_lst)
```
I don't know how to combine the Symint path and the Sympy.Expr path because (1) I'm not sure how to evaluate Sympy.Expr and (2) I need to guard_size_oblivious on the SymInts

zou3519 replied:
It is kind of annoying to combine the two paths because we would need to pass around V.graph (the GraphLowering?) and this needs to be threaded through a bunch of APIs

eellison replied:
Sorry, refreshing my memory on guard_size_obivious.. Isn't this equivalent to sizevars.statically_known_leq / other apis ? 

Can you convert the symints to sympy expr and then use those apis ?

zou3519 replied:
converting the symints to sympy expr requires me to thread through sizevars and that's a bit of a refactor... but I can attempt that if we think it's worth it? Otherwise this PR as-is fixes all my problems

eellison replied:
Isn't it just `s.node.expr`

zou3519 replied:
hmmm, let me try this

zou3519 committed:
Update on "[inductor] make requires_stride_order more unbacked-symint-aware"


Previously, we tried to sort SymInt strides to determine the stride
order. This PR makes the sorting more unbacked symint aware: given a Tensor
with sizes (u0, u1, u2), it has strides (u1 * u2, u1, 1), which is
sortable under the guard_size_oblivious assumptions.

Test Plan:
- test case

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

zou3519 committed:
Update on "[inductor] make requires_stride_order more unbacked-symint-aware"


Previously, we tried to sort SymInt strides to determine the stride
order. This PR makes the sorting more unbacked symint aware: given a Tensor
with sizes (u0, u1, u2), it has strides (u1 * u2, u1, 1), which is
sortable under the guard_size_oblivious assumptions.

Test Plan:
- test case

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

zou3519 committed:
Update on "[inductor] make requires_stride_order more unbacked-symint-aware"


Previously, we tried to sort SymInt strides to determine the stride
order. This PR makes the sorting more unbacked symint aware: given a Tensor
with sizes (u0, u1, u2), it has strides (u1 * u2, u1, 1), which is
sortable under the guard_size_oblivious assumptions.

Test Plan:
- test case

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

zou3519 committed:
Update on "[inductor] make requires_stride_order more unbacked-symint-aware"


Previously, we tried to sort SymInt strides to determine the stride
order. This PR makes the sorting more unbacked symint aware: given a Tensor
with sizes (u0, u1, u2), it has strides (u1 * u2, u1, 1), which is
sortable under the guard_size_oblivious assumptions.

Test Plan:
- test case

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

zou3519 committed:
Update on "[inductor] make requires_stride_order more unbacked-symint-aware"


Previously, we tried to sort SymInt strides to determine the stride
order. This PR makes the sorting more unbacked symint aware: given a Tensor
with sizes (u0, u1, u2), it has strides (u1 * u2, u1, 1), which is
sortable under the guard_size_oblivious assumptions.

Test Plan:
- test case

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

zou3519 commented:
trying something out, will re-request review if it works

zou3519 committed:
Update on "[inductor] make requires_stride_order more unbacked-symint-aware"


Previously, we tried to sort SymInt strides to determine the stride
order. This PR makes the sorting more unbacked symint aware: given a Tensor
with sizes (u0, u1, u2), it has strides (u1 * u2, u1, 1), which is
sortable under the guard_size_oblivious assumptions.

Test Plan:
- test case

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

zou3519 committed:
Update on "[inductor] make requires_stride_order more unbacked-symint-aware"


Previously, we tried to sort SymInt strides to determine the stride
order. This PR makes the sorting more unbacked symint aware: given a Tensor
with sizes (u0, u1, u2), it has strides (u1 * u2, u1, 1), which is
sortable under the guard_size_oblivious assumptions.

Test Plan:
- test case

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

zou3519 commented:
Okay should be good now

eellison commented on torch/_inductor/utils.py:
```diff
@@ -856,6 +856,59 @@ def argsort(seq) -> List[int]:
+    # Strategy: convert all symints to sympy.Expr, then evaluate them.
+    seq_with_exprs = []
+    maybe_shape_env = None
+    for idx, maybe_symint in enumerate(seq):
```
why do we need to do this ? inductor should always have a shape env ?

zou3519 replied:
> Looks good - just had one question about where you are getting the non shape env case. I dont want to leave misleading impressions about inductor invariants in code

The problem is get_stride_order is called from both inside and outside of inductor code. Would you prefer me to create two versions of get_stride_order -- one for calling inside of inductor, and one for calling outside of inductor?

The only callsite not in inductor is [here](https://www.internalfb.com/code/fbsource/[662193f8e44289b217d421f47ff9b96573060081]/fbcode/caffe2/torch/_higher_order_ops/flex_attention.py?lines=63), so it's not too bad to duplicate the code.

eellison replied:
A couple options:

a) The callsite there can probably use this same logic from prims - would prefer that for not mixing up the symint/sympy expr :

```
    logical_to_physical_perm = compute_elementwise_output_logical_to_physical_perm(
        *tensors, _skip_checks=True
    )
    permuted_shape = apply_perm(shape, logical_to_physical_perm)  # to physical
```

b) Can we have a common api where we take in the shape env ? in inductor, should always be V.graph.sizevars.shape_env ? whereas in tracing should be in fake mode.

zou3519 replied:
Updated to have an API where we accept the shape_env

eellison commented on torch/_inductor/utils.py:
```diff
@@ -856,6 +856,59 @@ def argsort(seq) -> List[int]:
+        def evaluate(expr):
+            if isinstance(expr, bool):
+                return expr
+            if maybe_shape_env is not None:
+                return maybe_shape_env.evaluate_expr(expr, size_oblivious=True)
+            # Grab Inductor's global shape_env as a last resort.
+            # We don't always do this because this helper function is called
+            # from non-Inductor contexts (but only with ints or symints)
+            from .virtualized import V
+
+            return V.graph._shape_env.evaluate_expr(expr, size_oblivious=True)
```
If we had to do this every where, inductor would get really verbose. It should an invariant that there is alway a shape env. Even if you run inductor with static shapes, `sizevars.shape_env` will still not be None.

If there are contexts where there are no shape env (which we should fix, ideally..), maybe could factor this out to a utility elsewhere where inductor will invoke this ?

Logic here looks good but i dont want to leave misleading impressions about what the invariants are inside inductor.

eellison reviewed (commented):
Looks good - just had one question about where you are getting the non shape env case. I dont want to leave misleading impressions about inductor invariants in code

ezyang commented on torch/_inductor/utils.py:
```diff
@@ -856,6 +856,59 @@ def argsort(seq) -> List[int]:
+            return 1
+        if a[0] > b[0]:
+            return -1
+        return 0
```
My high level comment is whether or not you could have used an approach like this and then directly reuse argsort:

```
    # Checks that there exists a permutation of the strides s.t. the tensor would be contiguous
    # Sorts (length, stride) pairs by stride
    #   
    # This sort is done in a size-oblivious way, which helps if we do a
    # comparison like 2048*u0 > u0; we just want this to return True
    # (and not worry about what if u0 is zero).
    class K(NamedTuple):
        size: int
        stride: int

        def __lt__(self, other):
            return guard_size_oblivious(self.stride < other.stride)
        
        def __gt__(self, other):
            return guard_size_oblivious(self.stride > other.stride)

        def __le__(self, other):
            return guard_size_oblivious(self.stride <= other.stride)

        def __ge__(self, other):
            return guard_size_oblivious(self.stride >= other.stride)

        def __eq__(self, other):
            return guard_size_oblivious(self.stride == other.stride)

    lengths_and_strides = sorted(map(K, a.shape, a.stride()))
```

zou3519 replied:
The way Python allows people to extend sorting is by passing in their own `key` function to `sorted`. That's what we do in this PR

ezyang replied:
fine ig

zou3519 committed:
Update on "[inductor] make requires_stride_order more unbacked-symint-aware"


Previously, we tried to sort SymInt strides to determine the stride
order. This PR makes the sorting more unbacked symint aware: given a Tensor
with sizes (u0, u1, u2), it has strides (u1 * u2, u1, 1), which is
sortable under the guard_size_oblivious assumptions.

Test Plan:
- test case

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

zou3519 committed:
Update on "[inductor] make requires_stride_order more unbacked-symint-aware"


Previously, we tried to sort SymInt strides to determine the stride
order. This PR makes the sorting more unbacked symint aware: given a Tensor
with sizes (u0, u1, u2), it has strides (u1 * u2, u1, 1), which is
sortable under the guard_size_oblivious assumptions.

Test Plan:
- test case

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

zou3519 committed:
Update on "[inductor] make requires_stride_order more unbacked-symint-aware"


Previously, we tried to sort SymInt strides to determine the stride
order. This PR makes the sorting more unbacked symint aware: given a Tensor
with sizes (u0, u1, u2), it has strides (u1 * u2, u1, 1), which is
sortable under the guard_size_oblivious assumptions.

Test Plan:
- test case

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge
